src/tests/integration/monitor-tabs.test.ts

```diff
@@ -0,0 +1,200 @@
+    })
+
+    it('has no effect for active tabs of other windows', () => {
+      const { tabs } = mocks.getState()
+      expect(tabs.get(17)).to.have.property('active', true)
+    })
+  })
+
+  describe('chrome.tabs.onReplaced', () => {
+    it("replaces a tab (create and remove) and updates store's state", () => {
+      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
+      const [callback] = mocks.chrome.tabs.onCreated.addListener.firstCall.args
+
+      callback({
+        id: 11,
+        windowId: 1,
+        active: false,
+        url: 'https://foo.com/abc',
+      })
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(11)!
+
+      expect(tab.id).to.equal(11)
+      expect(tab.windowId).to.equal(1)
+      expect(tab.domain).to.equal('foo.com')
+      expect(tab.url).to.equal('https://foo.com/abc')
+
+      const [callbackOnRemove] = mocks.chrome.tabs.onRemoved.addListener.firstCall.args
+
+      const removedTabId = 18
+      const removeInfo = {
+        id: 18,
+        windowId: 1,
+        active: false,
+        url: 'https://new.com/abc',
+      }
+
+      callbackOnRemove(removedTabId, removeInfo)
+
+      const stateOnRemoved = mocks.getState()
+      const tabsNotRemoved = stateOnRemoved.tabs
+
+      expect(tabsNotRemoved.get(removedTabId)).to.equal(undefined)
+    })
+  })
+
+  describe('chrome.windows.onCreated', () => {
+    it("Adds a newly created window to store's state", () => {
+      const [callback] = mocks.chrome.windows.onCreated.addListener.firstCall.args
+
+      const win = { id: 4, focused: true }
+      callback(win)
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onCreated') {
+        throw new Error('Expect chrome.windows.onCreated to have been the most recent action')
+      } else {
+        expect(state.mostRecentAction.payload).to.eql({ win })
+      }
+    })
+  })
+
+  describe('chrome.windows.onRemoved', () => {
+    it("Deletes a removed window from store's state", () => {
+      const [callback] = mocks.chrome.windows.onRemoved.addListener.firstCall.args
+      const removedWindow = { id: 1, focused: false }
+      callback(removedWindow.id)
+
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onRemoved') {
+        throw new Error('Expect chrome.windows.onRemvoed to have been the most recent action')
+      } else {
+        expect(state.mostRecentAction.payload.windowId).to.eql(removedWindow.id)
+      }
+    })
+  })
+
+  describe('chrome.windows.onFocusChanged', () => {
+    it("Changes windows focus status in store's state", () => {
+      const [callback] = mocks.chrome.windows.onFocusChanged.addListener.firstCall.args
+      const changedWindow = { id: 3, focused: true }
+      callback(changedWindow.id)
+      const state = mocks.getState()
+
+      if (state.mostRecentAction.type !== 'chrome.windows.onFocusChanged') {
```

Comment:
Similar note as above - we're trying to test the specific behavior of the responder, not the generic behavior that sets the mostRecentAction

Reply:
expect(state.focusedWindowId).to.equal(changedWindow.id)